_posts/2020-03-06-testing-components-vue-conf-2020.md

```diff
@@ -0,0 +1,133 @@
+
+```js
+import { mount } from 'vue/test-utils'
+import ScoreBoard from './ScoreBoard.vue'
+
+it('increases score', () => {
+  const wrapper = mount(ScoreBoard)
+  expect(wrapper.vm.amount).toBe(0)
+
+  const increase = wrapper.find('button.increase')
+  wrapper.trigger('click')
+  
+  expect(wrapper.vm.amount).toBe(1)
+})
+```
+
+This example interactions more with the UI in the wrapper, but if your components grow there could be much more required setups in your tests. There is a trade off on how much you want or can invest in writing tests. It might be tempting to test everything, but if you have a large code base this can easily become overwhelming. I took the Application Testing in Vue.js from Vueconf US in Austin this year from [Alex](https://twitter.com/hootlex) and [Rolf](https://twitter.com/rahaug) and learned a lot. They're the founders of [vueschool.io](https://vueschool.io/) and I would highly recommend checking out their site. One of the main sections that stuck out to me in this situation is using the `setData` function in favor of testing all of the user interactions. We have some larger and complex components here at Ritter and testing creates some long files. Let create another component that is a little more complex and show how to use the `setData` to improve our tests and make them easier to read and maintain.
+
+```js
+<template>
+  <div>
+    <span>{{ amount }}</span>
+    <span click="toggleDetails">Click me!</span>
+    <div v-if="showDetails">
+      <button @click="increase" class="increase">increase</button>
+      <button @click="decrease" >Decrease</button>
+    </div>
+  </div>
+</template>
+export default {
+  data() {
+    return {
+      showDetails: false,
+      amount: 0
+    }
+  },
+  methods: {
+    toggleDetails() {
+      this.showDetails = !this.showDetails
+    },
+    increase() {
+      this.amount++
+    },
+    decrease() {
+      this.amount--
+    }
+  }
+}
+```
+
+Above is a slightly more complex component that requires a little more user interaction to increase or decrease the amount. We could easily find the `Click me!` span and trigger another click event, but this is where the test complexity could build in larger components. Our goal is to stay close to unit testing and focus on one section, so using the `setData` function helps us with the. Let's refactor our test to handle this.
+
+```js
+import { mount } from 'vue/test-utils'
+import ScoreBoard from './ScoreBoard.vue'
+
+it('increases score', () => {
+  const wrapper = mount(ScoreBoard)
+  expect(wrapper.vm.amount).toBe(0)
+
+  wrapper.setData({
+    showDetails: true
+  })
+
+  wrapper.vm.increase()
+  
+  expect(wrapper.vm.amount).toBe(1)
+})
+```
```

Comment:
Want to add a summary or closing remark?

Reply:
Add a conclusion, thanks!